### PR TITLE
Apply filters to log subscriptions

### PR DESCRIPTION
**Pull request by joeldudley**
### Why is this change needed?

As per the Geth API, we want the ability to provide filters that limit the logs that are returned.

### What changes were made as part of this PR:

- Provide a high level list of the changes made
- List key areas for the reviewer 

### :rotating_light: Definition of done :rotating_light:
- [ ] Unit tests added to cover new or changed functionality 
- [ ] Docs pages updated to cover new or changed functionality
- [ ] [Changelog.md](https://github.com/obscuronet/go-obscuro/blob/main/docs/testnet/changelog.md) updated

### CODE REVIEW

File: go/obsclient/authclient.go [hunk 1]
```diff
@@ -96,7 +97,7 @@ func (ac *AuthObsClient) BalanceAt(ctx context.Context, blockNumber *big.Int) (*
 	return hexutil.DecodeBig(result)
 }
 
-func (ac *AuthObsClient) SubscribeFilterLogs(ctx context.Context, filterQuery ethereum.FilterQuery, ch chan types.Log) (ethereum.Subscription, error) {
+func (ac *AuthObsClient) SubscribeFilterLogs(ctx context.Context, filterQuery filters.FilterCriteria, ch chan types.Log) (ethereum.Subscription, error) {
```
Comment (joeldudley) [inline go/obsclient/authclient.go:100]: `filters.FilterCriteria` is an alias for `ethereum.FilterQuery`, but Geth uses the former on RPC APIs, so I wanted to hew to that.

File: go/rpc/encrypted_client.go [hunk 2]
```diff
@@ -133,11 +133,6 @@ func (c *EncRPCClient) Subscribe(ctx context.Context, namespace string, ch inter
 		return nil, fmt.Errorf("only subscriptions of type %s are supported", RPCSubscriptionTypeLogs)
 	}
 
-	logCh, ok := ch.(chan types.Log)
```
Comment (joeldudley) [inline go/rpc/encrypted_client.go:136]: Moving this block down a bit, original location didn't make sense in the flow.

File: go/rpc/encrypted_client.go [hunk 3]
```diff
@@ -201,18 +200,10 @@ func (c *EncRPCClient) createAuthenticatedLogSubscription(args []interface{}) (*
 	if len(args) < 2 {
 		logSubscription.Filter = &filters.FilterCriteria{}
 	} else {
-		// We marshal the filter criteria from a map to JSON, then back from JSON into a FilterCriteria.
```
Comment (joeldudley) [inline go/rpc/encrypted_client.go:204]: Don't know why we used to have this craziness, but seems to work now.

File: go/obsclient/authclient.go [hunk 4]
```diff
@@ -96,8 +106,15 @@ func (ac *AuthObsClient) BalanceAt(ctx context.Context, blockNumber *big.Int) (*
 	return hexutil.DecodeBig(result)
 }
 
-func (ac *AuthObsClient) SubscribeFilterLogs(ctx context.Context, filterQuery ethereum.FilterQuery, ch chan types.Log) (ethereum.Subscription, error) {
-	return ac.rpcClient.Subscribe(ctx, rpc.RPCSubscribeNamespace, ch, rpc.RPCSubscriptionTypeLogs, filterQuery)
+func (ac *AuthObsClient) SubscribeFilterLogs(ctx context.Context, filterCriteria filters.FilterCriteria, ch chan types.Log) (ethereum.Subscription, error) {
+	filterCriteriaMap := map[string]interface{}{
```
Comment (joeldudley) [inline go/obsclient/authclient.go:110]: We need this as a map, not a filter criteria object, to pass to EncRPCClient.

File: go/rpc/encrypted_client.go [hunk 5]
```diff
@@ -201,7 +200,9 @@ func (c *EncRPCClient) createAuthenticatedLogSubscription(args []interface{}) (*
...
+		// We marshal the filter criteria from a map to JSON, then back from JSON into a FilterCriteria. This is
```
Comment (joeldudley) [inline go/rpc/encrypted_client.go:203]: I expanded this comment a bit to explain why the round-trip here is necessary.

File: go/enclave/events/subscription_manager.go [hunk 6]
```diff
@@ -179,3 +182,57 @@ func isRelevant(userAddrs []string, account *gethcommon.Address) bool {
 
 	return false
 }
+
+// Applies `filterLogs`, below, to determine whether the log should be filtered out.
```
Comment (BedrockSquirrel) [inline go/enclave/events/subscription_manager.go:186]: Maybe make it clearer here that these are from the user. Like user-submitted filter or subscription filter or something.

**Review by BedrockSquirrel — APPROVED** [review]
LGTM

**Pull request merged**